File: backend/services/filters.py
"""Pool filtering and player options — ported from Streamlit app.py."""

from __future__ import annotations

import functools
from typing import Any

import nationality_groups as ng
import player_profiles as pp
import transfermarkt_profiles as tm
from position_families import (
    DEFAULT_POSITION_FAMILY,
    EUROPEAN_POSITION_FAMILIES,
    normalize_position_family,
    position_codes_for_family,
    position_family_label,
    rating_groups_for_family,
)

VALUE_SLIDER_MAX_EUR = 150_000_000
CONTRACT_YEAR_MIN = 2026
CONTRACT_YEAR_MAX = 2033
MINUTES_MIN = 0
MINUTES_MAX = 3600
HEIGHT_MIN_M = 1.60
HEIGHT_MAX_M = 2.05

MIDFIELD_POSITION_BLOCKS: tuple[tuple[str, str, frozenset[str] | None, str | None], ...] = (
    ("cm", "Meio-campistas centrais", None, "central_midfielders"),
    ("am", "Meio-campistas ofensivos", None, "attacking_midfielders"),
)

POSITION_FAMILY_OPTIONS = list(EUROPEAN_POSITION_FAMILIES)

PASS_SCORE_LETTER_FIELDS: dict[str, str] = {
    "volume_grade": "pass_volume_letter",
    "efficiency_grade": "pass_efficiency_letter",
    "buildup_grade": "pass_buildup_letter",
    "chance_grade": "pass_chance_creation_letter",
}

LETTER_GRADE_OPTIONS = [
    ("all", "Todas"),
    ("A+", "A+"),
    ("A", "A"),
    ("A-", "A−"),
    ("B+", "B+"),
    ("B", "B"),
    ("B-", "B−"),
    ("C+", "C+"),
    ("C", "C"),
    ("C-", "C−"),
    ("D", "D"),
]

LEAGUE_OPTIONS = [
    ("all", "All leagues"),
    ("premier_league", "Premier League"),
    ("italia_seriea", "Serie A"),
    ("laliga", "La Liga"),
    ("bundesliga", "Bundesliga"),
    ("ligue1", "Ligue 1"),
]

LEAGUE_LABEL_TO_KEY = {label.lower(): key for key, label in LEAGUE_OPTIONS if key != "all"}
LEAGUE_LABEL_TO_KEY.update({key: key for key, _label in LEAGUE_OPTIONS if key != "all"})


def normalize_league_filter_key(player: dict) -> str:
    source = str(player.get("league_source") or "").strip().lower()
    if source:
        return source
    label = str(player.get("league") or "").strip().lower()
    return LEAGUE_LABEL_TO_KEY.get(label, label)

FOOT_OPTIONS = [
    ("all", "Todos"),
    ("left", "Esquerdo"),
    ("right", "Direito"),
    ("both", "Ambidestro"),
]

AGE_BAND_OPTIONS: tuple[tuple[str, int | None, int | None], ...] = (
    ("all", None, None),
    ("u21", None, 21),
    ("u23", 22, 23),
    ("24_30", 24, 30),
    ("over30", 31, None),
)

AGE_BAND_LABELS = {
    "all": "Todas as idades",
    "u21": "U21",
    "u23": "U23",
    "24_30": "24-30",
    "over30": ">30",
}


def parse_age_band(age_band: str | None) -> tuple[int | None, int | None]:
    key = (age_band or "all").strip().lower()
    for band_key, lo, hi in AGE_BAND_OPTIONS:
        if band_key == key:
            return lo, hi
    return None, None


def position_blocks_for_family(position_family: str = DEFAULT_POSITION_FAMILY) -> list[tuple[str, str]]:
    family = normalize_position_family(position_family)
    family_label = position_family_label(family).lower()
    options: list[tuple[str, str]] = [("all", f"Todos os {family_label}")]
    if family == "midfielders":
        options.extend((block_id, label) for block_id, label, *_rest in MIDFIELD_POSITION_BLOCKS)
    return options


def filter_options_meta(position_family: str = DEFAULT_POSITION_FAMILY) -> dict[str, Any]:
    import nationality_groups as ng

    family = normalize_position_family(position_family)
    return {
        "position_families": [{"key": k, "label": l} for k, l in POSITION_FAMILY_OPTIONS],
        "leagues": [{"key": k, "label": l} for k, l in LEAGUE_OPTIONS],
        "foot": [{"key": k, "label": l} for k, l in FOOT_OPTIONS],
        "age_bands": [
            {"key": k, "label": AGE_BAND_LABELS.get(k, k), "min": lo, "max": hi}
            for k, lo, hi in AGE_BAND_OPTIONS
        ],
        "nationality_regions": list(ng.NATIONALITY_REGION_OPTIONS),
        "age_range": {"min": pp.MIN_PLAYER_AGE, "max": pp.MAX_PLAYER_AGE},
        "value_range_m": {"min": 0, "max": int(VALUE_SLIDER_MAX_EUR / 1_000_000)},
        "contract_year_range": {"min": CONTRACT_YEAR_MIN, "max": CONTRACT_YEAR_MAX},
        "minutes_range": {"min": MINUTES_MIN, "max": MINUTES_MAX},
        "height_range_m": {"min": HEIGHT_MIN_M, "max": HEIGHT_MAX_M},
        "letter_grades": [{"key": k, "label": l} for k, l in LETTER_GRADE_OPTIONS],
        "pass_score_filters": [
            {"key": "volume_grade", "label": "Volume"},
            {"key": "efficiency_grade", "label": "Efficiency"},
            {"key": "buildup_grade", "label": "Build-up"},
            {"key": "chance_grade", "label": "Chance creation"},
        ],
        "position_blocks": [{"key": k, "label": l} for k, l in position_blocks_for_family(family)],
        "defaults": {
            "league": "all",
            "position_family": family,
            "position_block": "all",
            "age_band": "all",
            "age_slider": [pp.MIN_PLAYER_AGE, pp.MAX_PLAYER_AGE],
            "foot": "all",
            "value_slider_m": [0, int(VALUE_SLIDER_MAX_EUR / 1_000_000)],
            "contract_year": [CONTRACT_YEAR_MIN, CONTRACT_YEAR_MAX],
            "minutes_slider": [MINUTES_MIN, MINUTES_MAX],
            "height_slider_m": [HEIGHT_MIN_M, HEIGHT_MAX_M],
            "nationality_regions": [ng.NATIONALITY_REGION_WORLD],
            "nationality_countries": [],
        },
    }


def normalize_filter_foot(value: str | None) -> str | None:
    if not value:
        return None
    text = str(value).strip().lower()
    if text in {"left", "esquerdo"}:
        return "left"
    if text in {"right", "direito"}:
        return "right"
    if text in {"both", "ambidestro"}:
        return "both"
    return None


def all_position_filters(position_family: str = DEFAULT_POSITION_FAMILY) -> tuple[frozenset[str], frozenset[str]]:
    family = normalize_position_family(position_family)
    return position_codes_for_family(family), rating_groups_for_family(family)


def position_filter_from_block(
    position_block: str = "all",
    *,
    position_family: str = DEFAULT_POSITION_FAMILY,
) -> tuple[frozenset[str], frozenset[str]]:
    family = normalize_position_family(position_family)
    block_id = (position_block or "all").strip().lower()
    if family != "midfielders" or block_id in {"", "all"}:
        return all_position_filters(family)
    for bid, _label, block_codes, rating_group in MIDFIELD_POSITION_BLOCKS:
        if bid != block_id:
            continue
        codes: set[str] = set()
        groups: set[str] = set()
        if block_codes:
            codes.update(block_codes)
        if rating_group:
            groups.add(rating_group)
        return frozenset(codes), frozenset(groups)
    return all_position_filters(family)


def player_matches_position_filter(
    player: dict,
    *,
    position_codes: frozenset[str],
    position_groups: frozenset[str],
) -> bool:
    if not position_codes and not position_groups:
        return True
    pos = str(player.get("position") or "").strip().upper()
    group = str(player.get("position_group") or "")
    if position_groups and group in position_groups:
        return True
    if position_codes and pos in position_codes:
        return True
    return False


def filter_player_pool(
    all_players: list[dict],
    progression_by_id: dict[str, dict],
    *,
    league: str = "all",
    age_min: int | None = None,
    age_max: int | None = None,
    age_slider_min: int | None = None,
    age_slider_max: int | None = None,
    foot: str = "all",
    value_min_eur: int = 0,
    value_max_eur: int = VALUE_SLIDER_MAX_EUR,
    contract_year_min: int = CONTRACT_YEAR_MIN,
    contract_year_max: int = CONTRACT_YEAR_MAX,
    minutes_min: int = MINUTES_MIN,
    minutes_max: int = MINUTES_MAX,
    height_min_m: float = HEIGHT_MIN_M,
    height_max_m: float = HEIGHT_MAX_M,
    nationalities: list[str] | None = None,
) -> list[dict]:
    effective_age_min = max(age_min or pp.MIN_PLAYER_AGE, age_slider_min or pp.MIN_PLAYER_AGE)
    effective_age_max = min(age_max or pp.MAX_PLAYER_AGE, age_slider_max or pp.MAX_PLAYER_AGE)
    filter_by_value = value_min_eur > 0 or value_max_eur < VALUE_SLIDER_MAX_EUR
    filter_by_contract = contract_year_min > CONTRACT_YEAR_MIN or contract_year_max < CONTRACT_YEAR_MAX
    filter_by_minutes = minutes_min > MINUTES_MIN or minutes_max < MINUTES_MAX
    filter_by_height = height_min_m > HEIGHT_MIN_M or height_max_m < HEIGHT_MAX_M
    allowed_nationalities = set(nationalities) if nationalities else None

    out: list[dict] = []
    for player in all_players:
        pid = str(player["player_id"])
        if league != "all" and normalize_league_filter_key(player) != league.lower():
            continue
        age = player.get("age")
        if age is None:
            age = pp.read_cached_age(pid)
        if age is not None:
            age_val = int(age)
            if age_val < effective_age_min or age_val > effective_age_max:
                continue
        elif age_min is not None or age_max is not None:
            continue
        if foot != "all":
            player_foot = normalize_filter_foot(
                player.get("dominant_foot") or pp.read_cached_dominant_foot(pid)
            )
            if player_foot is None or player_foot != foot:
                continue
        if filter_by_value:
            market_value_eur = player.get("market_value_eur")
            if market_value_eur is None:
                market_value_eur = tm.read_cached_market_value_eur(pid)
            if market_value_eur is None:
                continue
            mv = int(market_value_eur)
            if mv < value_min_eur or mv > value_max_eur:
                continue
        if filter_by_contract:
            contract_until = player.get("contract_until")
            if not contract_until:
                contract_until = pp.read_cached_profile(pid).get("contract_until")
            if not contract_until:
                continue
            try:
                contract_year = int(str(contract_until)[:4])
            except (TypeError, ValueError):
                continue
            if contract_year < contract_year_min or contract_year > contract_year_max:
                continue
        if filter_by_minutes:
            minutes = player.get("minutes")
            if minutes is None:
                continue
            try:
                minutes_val = int(minutes)
            except (TypeError, ValueError):
                continue
            if minutes_val < minutes_min or minutes_val > minutes_max:
                continue
        if filter_by_height:
            height_raw = player.get("height") or pp.read_cached_profile(pid).get("height")
            height_m = pp.parse_height_meters(height_raw)
            if height_m is None:
                continue
            if height_m < height_min_m or height_m > height_max_m:
                continue
        if allowed_nationalities is not None:
            nationality = player.get("nationality") or pp.read_cached_nationality(pid)
            if not ng.nationality_matches_filter(nationality, allowed=allowed_nationalities):
                continue
        out.append(player)
    return out


def _normalize_letter_grade(letter: str | None) -> str:
    return str(letter or "").strip().upper().replace("−", "-")


def _letter_grade_score(letter: str | None) -> float | None:
    from xp_stats_engine import LETTER_GRADE_COLOR_SCORES

    normalized = _normalize_letter_grade(letter)
    if not normalized or normalized == "—":
        return None
    return LETTER_GRADE_COLOR_SCORES.get(normalized)


def letter_grade_meets_minimum(player_letter: str | None, minimum_letter: str) -> bool:
    """True when the player letter is at least the selected minimum grade."""
    min_score = _letter_grade_score(minimum_letter)
    if min_score is None:
        return True
    player_score = _letter_grade_score(player_letter)
    if player_score is None:
        return False
    return player_score >= min_score


def matches_pass_letter_filters(
    xp_profile: dict,
    *,
    volume_grade: str = "all",
    efficiency_grade: str = "all",
    buildup_grade: str = "all",
    chance_grade: str = "all",
) -> bool:
    checks = {
        "volume_grade": volume_grade,
        "efficiency_grade": efficiency_grade,
        "buildup_grade": buildup_grade,
        "chance_grade": chance_grade,
    }
    for param_key, selected in checks.items():
        if not selected or selected == "all":
            continue
        letter_key = PASS_SCORE_LETTER_FIELDS[param_key]
        player_letter = xp_profile.get(letter_key)
        if not letter_grade_meets_minimum(player_letter, selected):
            return False
    return True


def filter_players_by_pass_letters(
    players: list[dict],
    xp_by_id: dict[str, dict],
    *,
    volume_grade: str = "all",
    efficiency_grade: str = "all",
    buildup_grade: str = "all",
    chance_grade: str = "all",
) -> list[dict]:
    if all(g == "all" for g in (volume_grade, efficiency_grade, buildup_grade, chance_grade)):
        return players
    out: list[dict] = []
    for player in players:
        pid = str(player["player_id"])
        xp_profile = xp_by_id.get(pid, {})
        if matches_pass_letter_filters(
            xp_profile,
            volume_grade=volume_grade,
            efficiency_grade=efficiency_grade,
            buildup_grade=buildup_grade,
            chance_grade=chance_grade,
        ):
            out.append(player)
    return out


def fmt_rating_score(value: Any) -> str:
    if value is None:
        return "—"
    try:
        return f"{float(value):.1f}"
    except (TypeError, ValueError):
        return "—"


def player_options(
    players: list[dict],
    progression_by_id: dict[str, dict],
    *,
    xp_by_id: dict[str, dict] | None = None,
    exclude_player_id: str | None = None,
    sort_by: str = "xp_pass_rating",
    position_block: str = "all",
    position_family: str = DEFAULT_POSITION_FAMILY,
) -> list[dict[str, str]]:
    position_codes, position_groups = position_filter_from_block(
        position_block,
        position_family=position_family,
    )
    ranked_rows: list[tuple[str, str, str, float]] = []
    for player in players:
        pid = str(player["player_id"])
        if exclude_player_id and pid == str(exclude_player_id):
            continue
        profile = progression_by_id.get(pid, player)
        if not player_matches_position_filter(
            profile,
            position_codes=position_codes,
            position_groups=position_groups,
        ):
            continue
        xp_profile = (xp_by_id or {}).get(pid, {})
        if sort_by == "xp_pass_rating":
            rating_val = xp_profile.get("xp_pass_rating")
            sort_key = float(rating_val) if rating_val is not None else float("-inf")
        else:
            sort_key = float(xp_profile.get("xp_m4_total", 0.0))
        ranked_rows.append((
            pid,
            str(player.get("player_name", "—")),
            str(player.get("team", "—")),
            sort_key,
        ))

    ranked_rows.sort(key=lambda row: (-row[3], row[1].lower()))
    options: list[dict[str, str]] = []
    for idx, (pid, name, team, sort_key) in enumerate(ranked_rows, start=1):
        xp_profile = (xp_by_id or {}).get(pid, {})
        if sort_by == "xp_pass_rating":
            rating_val = xp_profile.get("xp_pass_rating")
            if rating_val is not None:
                try:
                    suffix = f"· {float(rating_val) * 10:.1f}"
                except (TypeError, ValueError):
                    suffix = "· —"
            else:
                suffix = "· —"
        else:
            suffix = f"· xP {sort_key:.1f}"
        options.append({
            "player_id": pid,
            "player_name": name,
            "team": team,
            "label": f"#{idx} {name} ({team}) {suffix}",
        })
    return options


def available_nationalities(all_players: list[dict]) -> list[str]:
    nationalities: set[str] = set()
    for player in all_players:
        pid = str(player.get("player_id", ""))
        nationality = player.get("nationality") or pp.read_cached_nationality(pid)
        normalized = ng.normalize_nationality(nationality)
        if normalized:
            nationalities.add(normalized)
    return sorted(nationalities)

File: frontend/lib/static/filters.ts
import type { ProfileFilterState } from "@/lib/profileParams";
import { filtersToApiParams } from "@/lib/profileParams";
import { positionBlocksForFamily } from "@/lib/positionFamilies";
import {
  LETTER_GRADE_SCORES,
  PASS_SCORE_LETTER_FIELDS,
  PLAYER_LIST_FIELDS,
} from "@/lib/static/constants";
import type { PoolPlayer } from "@/lib/static/pool";
import { loadNationalityRegions } from "@/lib/static/pool";

const VALUE_SLIDER_MAX_EUR = 150_000_000;
const MIN_AGE = 16;
const MAX_AGE = 42;

type AgeBand = { key: string; min: number | null; max: number | null };

const AGE_BANDS: AgeBand[] = [
  { key: "all", min: null, max: null },
  { key: "u21", min: null, max: 21 },
  { key: "u23", min: 22, max: 23 },
  { key: "24_30", min: 24, max: 30 },
  { key: "over30", min: 31, max: null },
];

function parseAgeBand(ageBand?: string): { min: number | null; max: number | null } {
  const key = (ageBand || "all").trim().toLowerCase();
  const band = AGE_BANDS.find((b) => b.key === key) ?? AGE_BANDS[0];
  return { min: band.min, max: band.max };
}

function normalizeFoot(value: unknown): string | null {
  const text = String(value ?? "").trim().toLowerCase();
  if (text === "left" || text === "esquerdo") return "left";
  if (text === "right" || text === "direito") return "right";
  if (text === "both" || text === "ambidestro") return "both";
  return null;
}

function letterGradeScore(letter: unknown): number | null {
  const normalized = String(letter ?? "")
    .trim()
    .toUpperCase()
    .replace("−", "-");
  if (!normalized || normalized === "—") return null;
  return LETTER_GRADE_SCORES[normalized] ?? null;
}

function letterGradeMeetsMinimum(playerLetter: unknown, minimumLetter: string): boolean {
  const minScore = letterGradeScore(minimumLetter);
  if (minScore === null) return true;
  const playerScore = letterGradeScore(playerLetter);
  if (playerScore === null) return false;
  return playerScore >= minScore;
}

function parseHeightMeters(value: unknown): number | null {
  if (value === null || value === undefined || value === "") return null;
  if (typeof value === "number" && Number.isFinite(value)) {
    return value > 3 ? value / 100 : value;
  }
  const text = String(value).trim().replace(",", ".");
  const m = text.match(/^(\d+(?:\.\d+)?)\s*m?$/i);
  if (m) {
    const n = parseFloat(m[1]);
    return n > 3 ? n / 100 : n;
  }
  const cm = text.match(/^(\d+)\s*cm$/i);
  if (cm) return parseInt(cm[1], 10) / 100;
  return null;
}

function positionFilterFromBlock(
  positionBlock: string,
  positionFamily: string,
): { codes: Set<string>; groups: Set<string> } {
  const blockId = (positionBlock || "all").trim().toLowerCase();
  const codes = new Set<string>();
  const groups = new Set<string>();
  if (positionFamily !== "midfielders" || blockId === "" || blockId === "all") {
    return { codes, groups };
  }
  if (blockId === "cm") groups.add("central_midfielders");
  if (blockId === "am") groups.add("attacking_midfielders");
  return { codes, groups };
}

function playerMatchesPositionFilter(
  player: PoolPlayer,
  codes: Set<string>,
  groups: Set<string>,
): boolean {
  if (codes.size === 0 && groups.size === 0) return true;
  const pos = String(player.position ?? "").trim().toUpperCase();
  const group = String(player.position_group ?? "");
  if (groups.size > 0 && groups.has(group)) return true;
  if (codes.size > 0 && codes.has(pos)) return true;
  return false;
}

async function resolveNationalities(filters: ProfileFilterState): Promise<Set<string> | null> {
  const regionsRaw = filters.regions?.split(",").map((r) => r.trim()).filter(Boolean) ?? ["World"];
  if (regionsRaw.length === 0 || regionsRaw.includes("World")) return null;

  const data = await loadNationalityRegions();
  const countriesByRegion = (data.countries_by_region ?? {}) as Record<string, string[]>;
  const aliases = (data.aliases ?? {}) as Record<string, string>;
  const allowed = new Set<string>();

  for (const region of regionsRaw) {
    for (const country of countriesByRegion[region] ?? []) {
      allowed.add(country);
      const alias = aliases[country.toLowerCase()];
      if (alias) allowed.add(alias);
    }
  }

  const countries = filters.countries?.split(",").map((c) => c.trim()).filter(Boolean) ?? [];
  if (countries.length > 0) {
    const narrowed = new Set<string>();
    for (const c of countries) narrowed.add(c);
    return narrowed;
  }

  return allowed.size > 0 ? allowed : null;
}

function nationalityMatches(nationality: unknown, allowed: Set<string>): boolean {
  const norm = String(nationality ?? "").trim();
  if (!norm) return false;
  if (allowed.has(norm)) return true;
  return [...allowed].some((c) => c.toLowerCase() === norm.toLowerCase());
}

export function pickFields(player: PoolPlayer, fields: readonly string[]): PoolPlayer {
  const out: PoolPlayer = {};
  for (const key of fields) {
    if (key in player) out[key] = player[key];
  }
  return out;
}

export async function filterPoolPlayers(
  players: PoolPlayer[],
  filters: ProfileFilterState = {},
): Promise<PoolPlayer[]> {
  const api = filtersToApiParams(filters);
  const { min: ageBandMin, max: ageBandMax } = parseAgeBand(api.age_band);
  const ageSliderMin = filters.age_min ? parseInt(filters.age_min, 10) : MIN_AGE;
  const ageSliderMax = filters.age_max ? parseInt(filters.age_max, 10) : MAX_AGE;
  const effectiveAgeMin = Math.max(ageBandMin ?? MIN_AGE, ageSliderMin);
  const effectiveAgeMax = Math.min(ageBandMax ?? MAX_AGE, ageSliderMax);

  const valueMinEur = parseInt(api.value_min_m ?? "0", 10) * 1_000_000;
  const valueMaxEur = parseInt(api.value_max_m ?? "150", 10) * 1_000_000;
  const contractMin = parseInt(api.contract_year_min ?? "2026", 10);
  const contractMax = parseInt(api.contract_year_max ?? "2033", 10);
  const minutesMin = parseInt(api.minutes_min ?? "0", 10);
  const minutesMax = parseInt(api.minutes_max ?? "3600", 10);
  const heightMin = parseFloat(api.height_min_m ?? "1.60");
  const heightMax = parseFloat(api.height_max_m ?? "2.05");

  const filterByValue = valueMinEur > 0 || valueMaxEur < VALUE_SLIDER_MAX_EUR;
  const filterByContract = contractMin > 2026 || contractMax < 2033;
  const filterByMinutes = minutesMin > 0 || minutesMax < 3600;
  const filterByHeight = heightMin > 1.6 || heightMax < 2.05;
  const allowedNationalities = await resolveNationalities(filters);

  const league = api.league || "all";
  const foot = api.foot || "all";
  const search = filters.search?.toLowerCase();

  let rows = players.filter((player) => {
    if (league !== "all" && String(player.league_source ?? "") !== league) return false;
    if (search && !String(player.player_name ?? "").toLowerCase().includes(search)) return false;

    const age = player.age;
    if (age !== null && age !== undefined && age !== "") {
      const ageVal = parseInt(String(age), 10);
      if (ageVal < effectiveAgeMin || ageVal > effectiveAgeMax) return false;
    } else if (ageBandMin !== null || ageBandMax !== null) {
      return false;
    }

    if (foot !== "all") {
      const playerFoot = normalizeFoot(player.dominant_foot);
      if (!playerFoot || playerFoot !== foot) return false;
    }

    if (filterByValue) {
      const mv = player.market_value_eur;
      if (mv === null || mv === undefined) return false;
      const val = parseInt(String(mv), 10);
      if (val < valueMinEur || val > valueMaxEur) return false;
    }

    if (filterByContract) {
      const contract = player.contract_until;
      if (!contract) return false;
      const year = parseInt(String(contract).slice(0, 4), 10);
      if (Number.isNaN(year) || year < contractMin || year > contractMax) return false;
    }

    if (filterByMinutes) {
      const minutes = player.minutes;
      if (minutes === null || minutes === undefined) return false;
      const val = parseInt(String(minutes), 10);
      if (val < minutesMin || val > minutesMax) return false;
    }

    if (filterByHeight) {
      const heightM = parseHeightMeters(player.height);
      if (heightM === null || heightM < heightMin || heightM > heightMax) return false;
    }

    if (allowedNationalities) {
      if (!nationalityMatches(player.nationality, allowedNationalities)) return false;
    }

    return true;
  });

  const gradeFilters: Record<string, string> = {
    volume_grade: api.volume_grade || "all",
    efficiency_grade: api.efficiency_grade || "all",
    buildup_grade: api.buildup_grade || "all",
    chance_grade: api.chance_grade || "all",
  };

  const hasGradeFilter = Object.values(gradeFilters).some((g) => g && g !== "all");
  if (hasGradeFilter) {
    rows = rows.filter((player) => {
      for (const [paramKey, selected] of Object.entries(gradeFilters)) {
        if (!selected || selected === "all") continue;
        const letterKey = PASS_SCORE_LETTER_FIELDS[paramKey];
        if (!letterGradeMeetsMinimum(player[letterKey], selected)) return false;
      }
      return true;
    });
  }

  return rows;
}

export function buildPlayerOptions(
  players: PoolPlayer[],
  filters: ProfileFilterState = {},
  excludePlayerId?: string,
): { player_id: string; player_name: string; team: string; label: string }[] {
  const family = filters.position_family ?? "midfielders";
  const { codes, groups } = positionFilterFromBlock(filters.position_block ?? "all", family);

  const ranked: { pid: string; name: string; team: string; sortKey: number }[] = [];
  for (const player of players) {
    const pid = String(player.player_id ?? "");
    if (!pid || (excludePlayerId && pid === excludePlayerId)) continue;
    if (!playerMatchesPositionFilter(player, codes, groups)) continue;
    const ratingVal = player.xp_pass_rating;
    const sortKey =
      ratingVal !== null && ratingVal !== undefined ? parseFloat(String(ratingVal)) : Number.NEGATIVE_INFINITY;
    ranked.push({
      pid,
      name: String(player.player_name ?? "—"),
      team: String(player.team ?? "—"),
      sortKey,
    });
  }

  ranked.sort((a, b) => b.sortKey - a.sortKey || a.name.localeCompare(b.name));
  return ranked.map((row, idx) => ({
    player_id: row.pid,
    player_name: row.name,
    team: row.team,
    label: `#${idx + 1} ${row.name} (${row.team}) · Pass ${row.sortKey > -Infinity ? row.sortKey.toFixed(1) : "—"}`,
  }));
}

export function listPlayerSummaries(
  players: PoolPlayer[],
  filters: ProfileFilterState,
  limit = 200,
  offset = 0,
): { total: number; players: PoolPlayer[] } {
  return {
    total: players.length,
    players: players.slice(offset, offset + limit).map((p) => pickFields(p, PLAYER_LIST_FIELDS)),
  };
}

export function availablePositionBlocks(family: string) {
  return positionBlocksForFamily(family);
}
